file-tailer/src/main/java/co/cask/cdap/filetailer/tailer/LogTailer.java

```
@@ -0,0 +1,422 @@
+/*
+ * Copyright 2014 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.filetailer.tailer;
+
+import co.cask.cdap.filetailer.AbstractWorker;
+import co.cask.cdap.filetailer.config.PipeConfiguration;
+import co.cask.cdap.filetailer.event.FileTailerEvent;
+import co.cask.cdap.filetailer.metrics.FileTailerMetricsProcessor;
+import co.cask.cdap.filetailer.queue.FileTailerQueue;
+import co.cask.cdap.filetailer.state.FileTailerState;
+import co.cask.cdap.filetailer.state.FileTailerStateProcessor;
+import co.cask.cdap.filetailer.state.exception.FileTailerStateProcessorException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.charset.Charset;
+import java.util.Comparator;
+import java.util.TreeMap;
+
+/**
+ * Tailer daemon
+ */
+public class LogTailer extends AbstractWorker {
+
+  private static final Logger LOG = LoggerFactory.getLogger(LogTailer.class);
+  private static final String RAF_MODE = "r";
+  private static final int DEFAULT_BUFSIZE = 4096;
+
+  private final long sleepInterval;
+  private final File logDirectory;
+  private final String logFileName;
+  private final String charsetName;
+  private Charset charset;
```

You can have this field as final and no need to have `charsetName`. If a given charset is not support, this class should fail early (meaning in constructor).